stackhpc_ipa_hardware_managers/system_nic.py

```diff
@@ -0,0 +1,367 @@
+
+def _is_nic_verification_disabled(node):
+    disable_check = node['extra'].get(_FIRMWARE_CHECK_DISABLE_KEY)
+    return strutils.bool_from_string(
+        disable_check) if disable_check is not None else False
+
+
+NICFirmwareVerifyResult = namedtuple(
+    'NICFirmwareVerifyResult',
+    'actual_version expected_version '
+    'matcher')
+
+NICDescriptor = namedtuple('NICDescriptor', "name pci_id firmware_version")
+
+
+class SystemNICHardwareManager(hardware.HardwareManager):
+    """Checks firmware version for a given network card"""
+
+    HARDWARE_MANAGER_VERSION = "1.0"
+
+    def evaluate_hardware_support(self):
+        """Declare whether the system is supported by this manager.
+
+        :returns: HardwareSupport level for this manager.
+        """
+        # This should work for anything which supports ethtool
+        return hardware.HardwareSupport.SERVICE_PROVIDER
+
+    def get_clean_steps(self, node, ports):
+        """Get a list of clean steps with priority.
+
+        :param node: The node object as provided by Ironic.
+        :param ports: Port objects as provided by Ironic.
+        :returns: A list of cleaning steps, as a list of dicts.
+        """
+        return [{'step': 'verify_nic_firmware',
+                 'priority': 90,
+                 'interface': 'deploy',
+                 'reboot_requested': False,
+                 'abortable': True}]
+
+    def get_interface_descriptors(self):
+        """Enumerate all interfaces as NicDescriptor objects
+
+        :return: list of NicDescriptor
+        """
+        # there might be multiple identical cards, we must check them all
+        uevent_mappings = _parse_uevent(_uevent_lines())
+        result = []
+        for interface, uevent_device_info in uevent_mappings.iteritems():
+            pci_id = _get_pci_id(uevent_device_info)
+            ethtool_output = _get_ethtool_output(interface)
+            firmware = _get_ethtool_field(ethtool_output, "firmware-version")
+            result.append(NICDescriptor(pci_id=pci_id, name=interface,
+                                        firmware_version=firmware))
+        return result
+
+    def verify_nic_firmware(self, node, ports):
+        """
+        Verify all firmware versions specified by the nic_firmware property.
+
+        The nic_firmware property should take the form of a json list of
+        firmware matching criteria. The firmware matching criteria should
+        be a dictionary containing the following required fields:
+
+        - vendor_id : numeric pci vendor id
+        - device_id : numeric pci device id
+        - firmware_version : expected firmware version
+
+        the values of all fields should be strings. An example is given below::
+
+            [
+                { "vendor_id": "15B3", "device_id": "1013",
+                  "firmware_version": "12.20.1010" }
+            ]
+
+        The nic_firmware property can be manually set on a given node, as shown
+        in this example::
```

Comment:
Please could we update the README in the root of the git repo so that it contains some instructions on how to use this.